pkg/manager/member/utils.go

```diff
@@ -148,15 +154,20 @@ func statefulSetEqual(new apps.StatefulSet, old apps.StatefulSet) bool {
 }
 
 // templateEqual compares the new podTemplateSpec's spec with old podTemplateSpec's last applied config
-func templateEqual(new corev1.PodTemplateSpec, old corev1.PodTemplateSpec) bool {
+func templateEqual(new *apps.StatefulSet, old *apps.StatefulSet) bool {
 	oldConfig := corev1.PodSpec{}
-	if lastAppliedConfig, ok := old.Annotations[LastAppliedConfigAnnotation]; ok {
+	lastAppliedConfig, ok := old.Annotations[LastAppliedPodTemplateAnnotation]
+	if !ok {
+		// fallback to old annotations
+		lastAppliedConfig, ok = old.Spec.Template.Annotations[LastAppliedConfigAnnotation]
+	}
+	if ok {
 		err := json.Unmarshal([]byte(lastAppliedConfig), &oldConfig)
 		if err != nil {
 			glog.Errorf("unmarshal PodTemplate: [%s/%s]'s applied config failed,error: %v", old.GetNamespace(), old.GetName(), err)
 			return false
 		}
-		return apiequality.Semantic.DeepEqual(oldConfig, new.Spec)
+		return apiequality.Semantic.DeepEqual(oldConfig, new.Spec.Template.Spec)
```

Comment:
when a new field with the default value (e.g. `DNSPolicy`) is added in client-side, this function will return false, because the old configuration does not have this field. `tkmm.tikvUpgrader.Upgrade` will be called. 

however, the stateful set controller will not upgrade pods because after defaulting in apiserver, the pod template spec does not change.

does this have problems?

Reply:
Yes, that's right